### ISSUE TITLE
questions

### ISSUE THREAD

**Issue by shijielaw** [Leo-Q-316/TDGNN#3]
您好，在学习贵作的过程中，我遇到了两个问题，希望能得到您的指教。
1. aggregators.py中MeanAggregator的time值应该怎么设置呢？是要设置得比数据集中最大的时间值稍大即可吗？
2. 负采样的时候，您是如何实现的呢？
期待您的回复，谢谢！

**Comment by Leo-Q-316** [comment]
1. 这个time是做预测的那些边所发生的时间，通过这个过滤掉发生在这条边时间之后的一些邻居信息，只考虑这个时间之前的邻居信息，并不是一个预设的值，具体的实现逻辑在model.py graphsage中的loss函数。
2. 我们的方式是对于每一个时间点来说都采样在当前时间点不存在的边样本，采样数量和当前时间点正样本相同。

**Comment by Leo-Q-316** [comment]
1. 这个time是做预测的那些边所发生的时间，通过这个过滤掉发生在这条边时间之后的一些邻居信息，只考虑这个时间之前的邻居信息，并不是一个预设的值，具体的实现逻辑在model.py graphsage中的loss函数。
2. 我们的方式是对于每一个时间点来说都随机采样在当前时间点不存在的边样本，采样数量和当前时间点正样本相同。
________________________________
发件人: Rykert ***@***.***>
发送时间: 2021年8月2日 9:52
收件人: Leo-Q-316/TDGNN ***@***.***>
抄送: Subscribed ***@***.***>
主题: [Leo-Q-316/TDGNN] questions (#3)


您好，在学习贵作的过程中，我遇到了两个问题，希望能得到您的指教。

  1.  aggregators.py中MeanAggregator的time值应该怎么设置呢？是要设置得比数据集中最大的时间值稍大即可吗？
  2.  负采样的时候，您是如何实现的呢？
期待您的回复，谢谢！

―

**Comment by DearRykert** [comment]
非常感谢您的回复！我现在还遇到的一些问题是：

（1）在计算出mask矩阵后，为什么还要对一些元素（我没有理解这些元素对应的是什么样的节点？是目标节点本身吗？）进行加1操作呢？【for i in range(0, len(nodes)):    mask[i, unique_nodes[nodes[i]]] = mask[i, unique_nodes[nodes[i]]] + 1；】

（2）mask矩阵的行归一化操作是不是必须的？【num_neigh = mask.sum(1, keepdim=True)     mask = mask.div(num_neigh)】。

再次感谢您的耐心回复，谢谢！

**Comment by Leo-Q-316** [comment]
> 非常感谢您的回复！我现在还遇到的一些问题是：
> 
> （1）在计算出mask矩阵后，为什么还要对一些元素（我没有理解这些元素对应的是什么样的节点？是目标节点本身吗？）进行加1操作呢？【for i in range(0, len(nodes)): mask[i, unique_nodes[nodes[i]]] = mask[i, unique_nodes[nodes[i]]] + 1；】
> 
> （2）mask矩阵的行归一化操作是不是必须的？【num_neigh = mask.sum(1, keepdim=True) mask = mask.div(num_neigh)】。
> 
> 再次感谢您的耐心回复，谢谢！

**Comment by Leo-Q-316** [comment]
1. 是的，是目标节点本身；
2. 归一化是必须的，不然会可能导致scale issues.

**Comment by DearRykert** [comment]
非常感谢您的回复。
学生仍在不断的学习过程中，遇到的问题有些多，实在是打扰您了。
我发现在程序中，最终经过softmax得出来的scores是一个1×2的tensor，我不明白的是predict_y每次为什么只取scores的两个元素中的第二个呢？【predict_y.append(scores.cpu().detach().numpy()[0][1])】

**Comment by DearRykert** [comment]
您好！模型的一个输入是节点的随机初始化特征，这些随机特征在runData方法中（model.py第132&133行）利用：features = nn.Embedding(len(feat_data), 1000)；features.weight = nn.Parameter(torch.FloatTensor(feat_data), requires_grad=False)。
（1）您在此设置的是requires_grad=False，这是否意味着features.weight不会随着模型的训练而更新呢？如果features.weight不更新是否会导致我们得不到节点的嵌入呢？
（2）若在此设置requires_grad=True，那么模型训练完以后，features.weight是否就是节点的最终嵌入呢？
非常感谢您的耐心阅读！
